pkg/controller/seed-controller-manager/mla/userprojectbinding_controller.go

```diff
@@ -0,0 +1,250 @@
+			Role:         string(role),
+		}
+		if status, err := r.grafanaClient.UpdateOrgUser(ctx, userRole, org.ID, user.ID); err != nil {
+			return reconcile.Result{}, fmt.Errorf("unable to update grafana user role: %w (status: %s, message: %s)", err, pointer.StringPtrDerefOr(status.Status, "no status"), pointer.StringPtrDerefOr(status.Message, "no message"))
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *userProjectBindingReconciler) getOrgByProjectID(ctx context.Context, projectID string) (grafanasdk.Org, error) {
+	project := &kubermaticv1.Project{}
+	if err := r.Get(ctx, types.NamespacedName{Name: projectID}, project); err != nil {
+		return grafanasdk.Org{}, fmt.Errorf("failed to get project: %w", err)
+	}
+
+	orgID, ok := project.GetAnnotations()[grafanaOrgAnnotationKey]
+	if !ok {
+		return grafanasdk.Org{}, fmt.Errorf("project should have grafana org annotation set")
+	}
+	id, err := strconv.ParseUint(orgID, 10, 32)
+	if err != nil {
+		return grafanasdk.Org{}, err
+	}
+	return r.grafanaClient.GetOrgById(ctx, uint(id))
+}
+
+func (r *userProjectBindingReconciler) handleDeletion(ctx context.Context, userProjectBinding *kubermaticv1.UserProjectBinding) error {
+	org, err := r.getOrgByProjectID(ctx, userProjectBinding.Spec.ProjectID)
+	if err != nil {
+		return err
+	}
+	user, err := r.getGrafanaOrgUser(ctx, org.ID, userProjectBinding.Spec.UserEmail)
+	if err != nil {
+		return fmt.Errorf("unable to get user : %w", err)
+	}
+	if user != nil {
+		status, err := r.grafanaClient.DeleteOrgUser(ctx, user.OrgId, user.ID)
+		if err != nil {
+			return fmt.Errorf("failed to delete org user: %w (status: %s, message: %s)", err, pointer.StringPtrDerefOr(status.Status, "no status"), pointer.StringPtrDerefOr(status.Message, "no message"))
+		}
+	}
+
+	kubernetes.RemoveFinalizer(userProjectBinding, mlaFinalizer)
+	if err := r.Update(ctx, userProjectBinding); err != nil {
+		return fmt.Errorf("updating UserProjectBinding: %w", err)
+	}
+
+	return nil
+}
+
+func (r *userProjectBindingReconciler) getGrafanaOrgUser(ctx context.Context, orgID uint, email string) (*grafanasdk.OrgUser, error) {
+	users, err := r.grafanaClient.GetOrgUsers(ctx, orgID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, user := range users {
+		if user.Email == email {
+			return &user, nil
+		}
+	}
+	return nil, nil
+}
+
+func (r *userProjectBindingReconciler) addGrafanaOrgUser(ctx context.Context, orgID uint, email, role string) (*grafanasdk.OrgUser, error) {
+	req, err := http.NewRequest("GET", r.grafanaURL+"/api/user", nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add(r.grafanaHeader, email)
+	resp, err := r.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	type response struct {
+		ID uint `json:"id"`
+	}
+	res := &response{}
+	defer resp.Body.Close()
+	decoder := json.NewDecoder(resp.Body)
+	if err := decoder.Decode(res); err != nil || res.ID == 0 {
+		return nil, fmt.Errorf("unable to decode response : %w", err)
+	}
```

Comment:
is this https://pkg.go.dev/github.com/kubermatic/grafanasdk#Client.GetActualUser call you've been looking for?

Reply:
nope :) actual user - is the user that currently authorized, but we don't have any users authorized here, we need to add new user using oauth here, there are no methods for that in the sdk

Reply:
But your code is doing `GET /api/user`, literally `GetActualUser()`. Did you meant to do something else instead of getting the current user?

Reply:
> @aborilov 
> oh, yeah, by bad, we need to get, but I need to set header, so again, can't use this method, right?

https://pkg.go.dev/github.com/kubermatic/grafanasdk#NewClient takes http.Client, which is sdk.DefaultHTTPClient, which, in turn a default http.Client!! No timeouts! superbad. So I suggest to construct our own client with our own http.RoundTripper (which will set appropriate header).

Reply:
hm, can try to do that. but can we make it so this header will be set only for one endpoint?

Reply:
Sure, why not. RoundTripper has all control over the http.Request

Reply:
but is it worth it?
and also from doc
> RoundTrip should not modify the request, except for consuming and closing the Request's Body.

Reply:
I set the timeout to 15 sec, what do you think, is it a good value?

Reply:
👍